Tuner.h

```diff
@@ -0,0 +1,135 @@
+#ifndef LIBCHESS_TUNER_H
+#define LIBCHESS_TUNER_H
+
+#include <iostream>
+#include <string>
+#include <vector>
+
+namespace libchess {
+
+class TunableParameter {
+  public:
+    TunableParameter(std::string name, int value) noexcept
+        : name_(std::move(name)), value_(value) {}
+
+    [[nodiscard]] const std::string& name() const { return name_; }
+    [[nodiscard]] int value() const { return value_; }
+
+    void set_value(int value) noexcept { value_ = value; }
+
+    [[nodiscard]] std::string to_str() const noexcept {
+        return name_ + ": " + std::to_string(value_);
+    }
+
+  private:
+    std::string name_;
+    int value_;
+};
+
+enum class Result { BLACK_WIN, DRAW, WHITE_WIN };
+
+template <class Position> class NormalizedResult {
+  public:
+    NormalizedResult(const Position& position, Result result) noexcept
+        : position_(std::move(position)) {
+        switch (result) {
+        case Result::BLACK_WIN:
+            value_ = 0.0;
+            break;
+        case Result::DRAW:
+            value_ = 0.5;
+            break;
+        case Result::WHITE_WIN:
+            value_ = 1.0;
+            break;
+        }
+    }
+
+    [[nodiscard]] const Position& position() const noexcept { return position_; }
+    [[nodiscard]] float value() const noexcept { return value_; }
+
+  private:
+    Position position_;
+    float value_;
+};
+
+template <class Position> class Tuner {
+  public:
+    Tuner(std::vector<NormalizedResult<Position>> normalized_results,
+          std::vector<TunableParameter> tunable_parameters,
+          std::function<int(const Position&, const std::vector<TunableParameter>&)> eval_function)
+        : normalized_results_(std::move(normalized_results)),
+          tunable_parameters_(std::move(tunable_parameters)),
+          eval_function_(std::move(eval_function)) {}
+
+    [[nodiscard]] const std::vector<TunableParameter>& tunable_parameters() const noexcept {
+        return tunable_parameters_;
+    }
+
+    [[nodiscard]] float error() const noexcept {
+        float sum =
+            std::accumulate(normalized_results_.cbegin(), normalized_results_.cend(), 0.0,
+                            [this](int, const NormalizedResult<Position>& normalized_result) {
+                                float normalized_eval = sigmoid(eval(normalized_result.position()));
+                                float err = normalized_result.value() - normalized_eval;
+                                return err * err;
+                            });
+        return sum / float(normalized_results_.size());
```

Comment:
sqrt(sum / etc) to get a percentage

Reply:
Sqrt will reduce the precision (I just took the formula from the wiki here)